presto-main/src/main/java/com/facebook/presto/server/protocol/ExecutingStatementResource.java

```diff
@@ -100,7 +101,15 @@ public void getQueryResults(
         Query query = queryProvider.getQuery(queryId, slug);
         ListenableFuture<Response> queryResultsFuture = transform(
                 query.waitForResults(token, uriInfo, proto, wait, targetResultSize),
-                results -> toResponse(query, results, compressionEnabled),
+                (results) -> {
+                    Tracer tracer = query.getTracer();
+                    tracer.addPoint("Sending result data.");
+                    Response response = toResponse(query, results, compressionEnabled);
+                    if (results.getNextUri() == null) {
+                        tracer.endTrace("All results sent, query finished.");
+                    }
```

Comment:
We will have a lot of such points per query. Is that intended?

Reply:
I am not sure. But will be good to have a discussion with you as you know this flow better. I thought this was a good checkpoint when query approaches to finishing. But let's discuss it.